### PR TITLE
add API logs to API endpoints

### PR DESCRIPTION
**Pull request by annyhe**
for testing on heroku, and feedback.

### CODE REVIEW

**Comment by annyhe** [comment]
@pallavi2209 @iamigo This is the foundation for the API logs: additions to many files will look similar to api/v1/helpers/verbs/doFind.js. Please leave feedback.

File: utils/apiLog.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2016, salesforce.com, inc.
```
Comment (pallavi2209) [inline utils/apiLog.js]: 2017

File: utils/apiLog.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      method: req.method,
+    };
+    /* extract user, token and ipaddress and update log object */
+    jwtUtil.getTokenDetailsFromToken(req)
+    .then((resObj) => {
+      logObject.user = resObj.username;
+      logObject.token = resObj.tokenname;
+
+      // in-place modification of logObject
+      mapApiResultsToLogObject(resultObj, logObject);
```
Comment (pallavi2209) [inline utils/apiLog.js]: I think we should create logs even if token is not provided. In that case, this code should be outside then.

File: utils/apiLog.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+function mapApiResultsToLogObject(resultObj, logObject) {
+  if (resultObj.reqStartTime) {
+    logObject.totalTime = `${Date.now() - resultObj.reqStartTime}ms`;
+  }
+
+  if (resultObj.dbTime) {
+    logObject.dbTime = `${resultObj.dbTime}ms`;
+  }
+
+  if (resultObj.recordCount) {
```
Comment (pallavi2209) [inline utils/apiLog.js]: Would be great to have some comments on what each parameter interprets.

File: tests/logging/jwtUtil.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2016, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/logging/jwtUtil.js
```
Comment (pallavi2209) [inline tests/logging/jwtUtil.js]: Do you think we should have this file in an independent utils folder instead of logging folder?

File: tests/logging/jwtUtil.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  after(tu.forceDeleteUser);
+
+  it('works with valid Authorization from cookie', () => {
+    const cookies = {
+      Authorization: token,
+    };
+
+    jwtUtil.getTokenDetailsFromToken(cookies).then((res) => {
+      expect(res.tokenname).to.be.defined;
+      expect(res.usernname).to.be.defined;
```
Comment (pallavi2209) [inline tests/logging/jwtUtil.js]: Can we test the values as well?

File: tests/logging/enableApiLog.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+const tu = require('../testUtils');
+const u = require('./utils');
+const constants = require('../../api/v1/constants');
+const Aspect = tu.db.Aspect;
+const path = '/v1/samples/upsert/bulk';
+
+/* replicated api bulk upsert basic tests with enableApiLogs enabled
+ to execute api activity logging and test that it does not interfere with
+ the existing code. */
+describe('api: POST ' + path, () => {
```
Comment (pallavi2209) [inline tests/logging/enableApiLog.js]: Do you think it will be better to have tests for all method types?

File: utils/apiLog.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2016, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * /utils/apiLog.js
```
Comment (pallavi2209) [inline utils/apiLog.js]: There is a file activityLog in utils. Do you think it is better to move apiLog to activityLog because api request is a one of the activity type (api, worker and realtime)? Feel free to modify function names if conflicting.

File: api/v1/helpers/verbs/utils.js [hunk 8]
```diff
@@ -680,7 +681,7 @@ function forbidden(next, modelName) {
 // ----------------------------------------------------------------------------
 
 module.exports = {
-
+  logAPI,
```
Comment (pallavi2209) [inline api/v1/helpers/verbs/utils.js]: Not sure if this a good practice. Files not requiring logAPI will be importing the logAPI code for no reason.

File: utils/apiLog.js (hunk 7 above)
Comment (annyhe) [inline utils/apiLog.js]: let's keep them separate for now: activityLog has generic functions for all loggers, while apiLog has specific API logging functions.

File: utils/apiLog.js (hunk 2 above)
Comment (annyhe) [inline utils/apiLog.js]: if no token is provided, wouldn't the request err? In that case nothing would be logged, and an error would be thrown instead (403). 
Which endpoints would be token free? The /login and maybe /register?

File: tests/logging/jwtUtil.js (hunk 4 above)
Comment (annyhe) [inline tests/logging/jwtUtil.js]: We can move it to another folder at a later time. Currently there isn't any reason to move it.

File: tests/logging/enableApiLog.js (hunk 6 above)
Comment (annyhe) [inline tests/logging/enableApiLog.js]: Good point, I will add more tests for other endpoints, as API logging will cover not only POST, but PUT, PATCH, DELETE.

File: api/v1/helpers/verbs/utils.js (hunk 8 above)
Comment (annyhe) [inline api/v1/helpers/verbs/utils.js]: only a few files would be free from logAPI: most files in api/v1/helpers/verbs with the name do*.js would include logAPI.

File: tests/logging/enableApiLog.js (hunk 6 above)
Comment (annyhe) [inline tests/logging/enableApiLog.js]: Actually there might be a better way of duplicating tests: I'll run the api tests with the enableAPILogs token set to true. This way we check for the same results, and the tests should still pass.

File: package.json [hunk 9]
```diff
@@ -36,6 +36,7 @@
     "start-clock": "node --max_old_space_size=1500 --nouse-idle-notification clock/index.js",
     "start-worker": "node --max_old_space_size=1500 --nouse-idle-notification worker/jobProcessor.js",
     "test-api": "mocha -R dot --recursive tests/api",
+    "test-api-log": "ENABLE_API_LOGS=true mocha -R dot --recursive tests/api",
```
Comment (annyhe) [inline package.json]: testing the API logger by running the API tests with the config variable on. If any tests fail here while passing without  the API logger, we know there's something wrong with the logger implementation.

File: utils/apiLog.js (hunk 2 above)
Comment (pallavi2209) [inline utils/apiLog.js]: So, if the tokens are disabled at some point, like now, we wont be able to get an idea of other parameters as well, like dbTime etc. I think logging and token use should be independent. If token is not provided, we can just say token=None.

File: utils/apiLog.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    // create api activity log object
+    const logObject = {
+      ipAddress: activityLogUtil.getIPAddrFromReq(req),
+      responseBytes: getSize(req.body),
+      uri: req.url,
+      method: req.method,
+    };
+
+    if (!featureToggles.isFeatureEnabled('enforceApiToken')) {
+      // log with no token or user info
```
Comment (annyhe) [inline utils/apiLog.js]: log regardless of whether api token enforcement.

**Comment by annyhe** [comment]
@pallavi2209 @iamigo Ready for more comments :)

File: config/toggles.js [hunk 11]
```diff
@@ -69,6 +69,10 @@ const longTermToggles = {
   enforceWritePermission: environmentVariableTrue(pe,
                                         'ENFORCE_WRITE_PERMISSION'),
 
+  // Enable api activity logging
+  enableApiLogs: environmentVariableTrue(pe,
```
Comment (iamigo) [inline config/toggles.js]: Can we use "ENABLE_API_ACTIVITY_LOGS" so it follows the same naming convention as the worker one?

File: package.json (hunk 9 above)
Comment (iamigo) [inline package.json]: I like the blanket coverage by doing it this way, but it also means these tests are not counted by coveralls. Can you think of another way of achieving this? We've added a test utility function called "toggleOverride" which lets us turn a feature on or off in a test's before/after function (as long as the feature doesn't require re-initializing top-level app server index.js).

File: utils/apiLog.js (hunk 7 above)
Comment (pallavi2209) [inline utils/apiLog.js]: Then maybe create a folder activityLogs and put all the logging files together there? Just trying to prevent having too many root level files (similar kind) in utils folder.

File: package.json (hunk 9 above)
Comment (annyhe) [inline package.json]: @iamigo Another way of testing would be to toggleOverride enable api logs for all test api tests, but how would that differ from what is happening here?
As for how running this line not being counted by coveralls, as long as coveralls counts the original tests/api, we don't need coveralls to count it again.

Comment (iamigo) [inline package.json]: OK @annyhe

File: utils/apiLog.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    } else {
+      // if enforcing API token,
+      // extract user, token to update log object
+      jwtUtil.getTokenDetailsFromToken(req)
+      .then((resObj) => {
+        logObject.user = resObj.username;
+        logObject.token = resObj.tokenname;
+
+        // log with the token
+        combineAndLog(resultObj, logObject);
```
Comment (pallavi2209) [inline utils/apiLog.js]: Why check enforceApiToken? Since you are not catching any error from getTokenDetailsFromToken, why not just call combineAndLog outside then? Would be cleaner code and no dependency on enforceApiToken

File: utils/apiLog.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+  activityLogUtil.printActivityLogString(logObject, 'api');
+}
+
+/**
+ * Combines input obj with values from the request object
+ * then logs
+ * @param {Object} req - the request object
+ * @param {Object} resultObj - Object with the rest of the fields to print
+ */
+function logAPI(req, resultObj) {
```
Comment (pallavi2209) [inline utils/apiLog.js]: Can we test this function?

File: utils/apiLog.js (hunk 12 above)
Comment (annyhe) [inline utils/apiLog.js]: good point. will change

**Comment by pallavi2209** [comment]
I don't see code for the API's not using verbs (like user tokens, authenticate, register, ssoconfig etc). Requested @annyhe to create another PR for that.

File: utils/apiLog.js (hunk 13 above)
Comment (annyhe) [inline utils/apiLog.js]: Regarding the test, because there is an async function involved, sinon would need a passed-in callback to check that the .then contents have been called
https://www.sitepoint.com/sinon-tutorial-javascript-testing-mocks-spies-stubs/

File: utils/apiLog.js (hunk 7 above)
Comment (annyhe) [inline utils/apiLog.js]: let's move it in another pull request

**Comment by annyhe** [comment]
after speaking with @pallavi2209 there are missing areas in api that are not being logged. Will do a separate pull request for that (ie. api/controllers/**.js)

**Comment by annyhe** [comment]
@iamigo Please review.

File: api/v1/helpers/verbs/doDelete.js [hunk 14]
```diff
@@ -50,6 +53,8 @@ function doDelete(req, res, next, props) {
 
     // when a resource is deleted, delete all its associations too
     u.deleteAllAssociations(o, assocNames);
+    resultObj.retval = o.dataValues;
+    u.logAPI(req, resultObj);
```
Comment (iamigo) [inline api/v1/helpers/verbs/doDelete.js]: I don't like the mix of calling this function both as u.logAPI(...) and as logAPI(...) (line 36) in the same module--it's too confusing to the reader. Please standardize and call it one way across all API modules.

File: config/toggles.js [hunk 15]
```diff
@@ -69,6 +69,10 @@ const longTermToggles = {
   enforceWritePermission: environmentVariableTrue(pe,
                                         'ENFORCE_WRITE_PERMISSION'),
 
+  // Enable api activity logging
+  enableApiWorkerLogs: environmentVariableTrue(pe,
```
Comment (iamigo) [inline config/toggles.js]: `enableApiActivityLogs`

File: api/v1/helpers/verbs/doDelete.js (hunk 14 above)
Comment (annyhe) [inline api/v1/helpers/verbs/doDelete.js]: u.logAPI and logAPI were different functions. Renamed logAPI to logAuditAPI to avoid confusion.

**Comment by annyhe** [comment]
@iamigo Ready for more comments :)

File: api/v1/helpers/verbs/doDelete.js (hunk 14 above)
Comment (iamigo) [inline api/v1/helpers/verbs/doDelete.js]: So just to be clear, "logAuditAPI" is the old audit log stuff that we will remove once the new activity logging is in place?

Comment (annyhe) [inline api/v1/helpers/verbs/doDelete.js]: yes, logAuditAPI is the old audit log. The new API logger is called logAPI.

File: utils/jwtUtil.js [hunk 16]
```diff
@@ -91,7 +91,9 @@ function getTokenDetailsFromTokenString(s) {
     if (s) {
       jwt.verify(s, env.tokenSecret, {}, (err, decodedData) => {
         if (err !== null || !decodedData) {
-          return reject(err);
+          return reject(new apiErrors.ForbiddenError({
```
Comment (annyhe) [inline utils/jwtUtil.js]: @iamigo I added the forbidden status code, to be consistent with the other return.
